File: algorithms/medium/integer_break.cpp
class Solution {
public:
  int integerBreak(int n) {
    vector<int> dp(59, 0);
    int ans = 1;
    dp[1] = 1;
    for (int i = 2; i <= n; i++) {
      for (int j = 1; j <= i / 2; j++) {
        array<int, 4> arr{
          dp[j] * dp[i - j],
          j * dp[i - j],
          dp[j] * (i - j),
          j * (i - j),
        };
        int tmp = *max_element(arr.begin(), arr.end());
        dp[i] = max(dp[i], tmp);
      }
      ans = max(ans, dp[i]);
    }
    return ans;
  }
};

File: algorithms/medium/integer_break.rs
impl Solution {
    pub fn integer_break(n: i32) -> i32 {
        let mut dp = vec![0usize; 59];
        let mut ans = 1;
        dp[1] = 1;
        for i in 2..=(n as usize) {
            for j in 1..=i / 2 {
                dp[i] = vec![
                    dp[j] * dp[i - j],
                    j * dp[i - j],
                    dp[j] * (i - j),
                    j * (i - j),
                ]
                .iter()
                .max()
                .cloned()
                .unwrap()
                .max(dp[i]);
            }
            ans = ans.max(dp[i]);
        }
        ans as _
    }
}
